### PR TITLE
[Maintenance] Migrating util.js and chartTheme.js to typescript

### PR DESCRIPTION
**Pull request by ghosetuhin**
Link to issue: https://github.com/actualbudget/actual/issues/1483

### Replaced util.js with util.ts
- Added data types when easily inferrable to the functions in the file
- Fixed lint errors which were created during changes

### Replaced chart-theme.js with chart-theme.ts
- Created types for all the different styles being used
- Assigned data types to all the variables in the file

### CODE REVIEW

File: packages/desktop-client/src/components/reports/chart-theme.ts [hunk 1]
```diff
@@ -14,15 +21,42 @@ const letterSpacing = 'normal';
...
-const baseLabelStyles = {
+type LabelStyles = {
+  fontFamily: string;
+  fontSize: number;
+  letterSpacing: string;
+  fill: string;
+  stroke: string;
+};
+
+const baseLabelStyles: LabelStyles = {
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/chart-theme.ts:32]: **💬 suggestion:** ‏same issue here
```suggestion
const baseLabelStyles = {
```

File: packages/desktop-client/src/components/reports/chart-theme.ts [hunk 2]
```diff
@@ -41,7 +75,82 @@ const axisBaseStyles = {
...
+    padding: {
+      top: number;
+      left: number;
+      right: number;
+      bottom: number;
+    };
+  };
+};
+
+export const chartTheme: ChartTheme = {
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/chart-theme.ts:153]: **💬 suggestion:** ‏and here
```suggestion
export const chartTheme = {
```

File: packages/desktop-client/src/components/reports/chart-theme.ts [hunk 3]
```diff
@@ -112,8 +221,12 @@ export const chartTheme = {
...
 };
 
-export function getColorScale(name) {
-  const scales = {
+type ColorScales = {
+  [key: string]: string[];
+};
+
+export function getColorScale(name: string): string[] {
+  const scales: ColorScales = {
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/chart-theme.ts:229]: **💬 suggestion:** ‏would be better to use an inline style here if we want to be super-strict. Although personally I wouldn't add a manual type to it. Up to you.
```suggestion
  const scales: Record<string, string[]> = {
```

File: packages/desktop-client/src/components/reports/util.ts [hunk 4]
```diff
@@ -13,23 +17,23 @@ export async function runAll(queries, cb) {
   cb(data);
 }
 
-export function index(data, field, mapper) {
+export function index(data, field: string, mapper?: (input) => unknown) {
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/util.ts:20]: **💬 suggestion:** ‏could be improved with generics
```suggestion
export function index<T, K extends keyof T>(data: T[], field: K, mapper?: (input: T[K]) => string) {
```

File: packages/desktop-client/src/components/reports/util.ts [hunk 5]
```diff
@@ -13,23 +17,23 @@ export async function runAll(queries, cb) {
...
   const result = {};
   data.forEach(item => {
     result[mapper ? mapper(item[field]) : item[field]] = item;
   });
   return result;
 }
 
-export function indexStack(data, fieldName, field) {
+export function indexStack(data, fieldName: string, field: string) {
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/util.ts:28]: **💬 suggestion:** ‏this one and also `indexCashFlow` (below) could be improved with generics too.

**Review by MatissJanis — CHANGES_REQUESTED** [review]
👋 Hi @ghosetuhin ! Thanks for your first contribution! I left a couple of comments that I hope won't be too tricky for you to solve. And once they are solved I'll be happy to merge this PR :)

File: packages/desktop-client/src/components/reports/chart-theme.ts [hunk 6]
```diff
@@ -1,6 +1,13 @@
...
 
-let colorFades = {
+type ColorFades = {
+  blueFadeStart: string;
+  blueFadeEnd: string;
+  redFadeStart: string;
+  redFadeEnd: string;
+};
+
+let colorFades: ColorFades = {
```
[the remark this answers is not in the document]
Comment (ghosetuhin) [reply packages/desktop-client/src/components/reports/chart-theme.ts]: Thanks for the suggestion! Made the changes!

File: packages/desktop-client/src/components/reports/chart-theme.ts (hunk 1 above)
Comment (ghosetuhin) [reply packages/desktop-client/src/components/reports/chart-theme.ts]: Removed it

File: packages/desktop-client/src/components/reports/chart-theme.ts [hunk 7]
```diff
@@ -14,15 +21,42 @@ const letterSpacing = 'normal';
...
+  ticks: {
+    fill: string;
+    size: number;
+    stroke: string;
+  };
+  axisLabel: LabelStyles;
+  tickLabels: LabelStyles;
+};
+
+const axisBaseStyles: AxisBaseStyles = {
```
[the remark this answers is not in the document]
Comment (ghosetuhin) [reply packages/desktop-client/src/components/reports/chart-theme.ts]: Removed it

File: packages/desktop-client/src/components/reports/chart-theme.ts (hunk 2 above)
Comment (ghosetuhin) [reply packages/desktop-client/src/components/reports/chart-theme.ts]: Removed it

File: packages/desktop-client/src/components/reports/chart-theme.ts (hunk 3 above)
Comment (ghosetuhin) [reply packages/desktop-client/src/components/reports/chart-theme.ts]: Implemented it inline. Thanks for the suggestion!

File: packages/desktop-client/src/components/reports/util.ts (hunk 4 above)
Comment (ghosetuhin) [reply packages/desktop-client/src/components/reports/util.ts]: Please take a look at this. I implemented it using generics this time

File: packages/desktop-client/src/components/reports/util.ts (hunk 5 above)
Comment (ghosetuhin) [reply packages/desktop-client/src/components/reports/util.ts]: I tried implementing it for indexCashFlow but it is not taking generic types quite well

File: packages/desktop-client/src/components/reports/util.ts (hunk 4 above)
Comment (ghosetuhin) [reply packages/desktop-client/src/components/reports/util.ts]: the generics I implemented using the way you suggested were failing the type checks for some reason with the error Argument of type 'string' is not assignable to parameter of type 'never'. I tried playing around with it but couldn't fix it. Had to finally resort to using string as the type for some fields

File: packages/desktop-client/src/components/reports/util.ts [hunk 8]
```diff
@@ -13,23 +17,24 @@ export async function runAll(queries, cb) {
   cb(data);
 }
 
-export function index(data, field, mapper) {
-  const result = {};
+export function index<T>(data: T[], field: string, mapper?: (input) => string) {
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/util.ts:20]: Was a pretty fun problem to solve. Here you go: https://gist.github.com/MatissJanis/307e49bbeec34f2975709e0ec3bc3100 :)

Comment (ghosetuhin) [reply packages/desktop-client/src/components/reports/util.ts]: Ah, I see. Thanks for this!

**Review by MatissJanis — APPROVED** [review]
Thanks!

**Pull request merged**